### DISCUSSION TITLE
ImportError using abi3 wheel on windows

### DISCUSSION THREAD

**Discussion by NicolasHanus** [PyO3/maturin#956] [Q&A] [answered]
Hello,
I have a project that generates an extension module through PyO3 with abi3 feature. I am building on Windows. The wheel builds fine, I can install it, but I cannot use it. From git bash:

$ python --version
Python 3.8.0

$ maturin --version
maturin 0.12.19

$ maturin build --release --no-sdist
≡ƒöù Found pyo3 bindings with abi3 support for Python ΓëÑ 3.7
≡ƒÉì Using CPython 3.8 at C:\dev\Python38\python.exe to generate to link bindings (With abi3, an interpreter is only required on windows)
    Finished release [optimized] target(s) in 0.07s
≡ƒôª Built wheel for abi3 Python ΓëÑ 3.7 to C:\Users\nhanus\dev\git\test_maturin\target\wheels\thermal-0.1.0-cp37-abi3-win_amd64.whl

$ pip --version
pip 22.1.2 from c:\users\nhanus\dev\git\test_maturin\.venv\lib\site-packages\pip (python 3.8)

$ pip install target/wheels/thermal-0.1.0-cp37-abi3-win_amd64.whl
Processing c:\users\nhanus\dev\git\test_maturin\target\wheels\thermal-0.1.0-cp37-abi3-win_amd64.whl
Installing collected packages: thermal
Successfully installed thermal-0.1.0

$ python
Python 3.8.0 (tags/v3.8.0:fa919fd, Oct 14 2019, 19:37:50) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import thermal
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\nhanus\dev\git\test_maturin\.venv\lib\site-packages\thermal\__init__.py", line 1, in <module>
    from .thermal import *
**ImportError: DLL load failed while importing thermal: The specified procedure could not be found.**

Here are the config files for my project:

Cargo.toml

```
[package]
name = "thermal"
version = "0.1.0"
authors = ["nico"]
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
name = "thermal"
# "cdylib" is necessary to produce a shared library for Python to import from.
crate-type = ["cdylib"]

[dependencies.pyo3]
version = "0.16.5"
features = ["extension-module", "abi3-py37"]
```

pyproject.toml

```
[build-system]
requires = ["maturin>=0.12,<0.13"]
build-backend = "maturin"
```

I've been struggling with that for some days now. I'm out of ideas.
Note that it works fine if I remove the 'abi3-py37' flag from the features, but I'd like to support multiple python versions.

Ultimately I'd like to cross compile this from Linux. So far, cross-compiling with the 'abi3-py37' flag, i manage to build the windows wheel but I have the same error at the end when importing it. Without the flag, I am not able to get a wheel (interpreter not found issue).

Can somebody please help ? Any idea what is going on ? What am I doing wrong ?

Many thanks in advance

**Comment by messense** [comment, marked as answer]
Run `python -c 'import sys; print(sys.base_prefix)'` then go to the printed directory, verify that it contains a `python3.dll` file.

Edit: solution is https://github.com/PyO3/maturin/discussions/956#discussioncomment-2922643

**Reply by NicolasHanus** [reply]
Thx for your answer. It seems to:
```
$ python -c 'import sys; print(sys.base_prefix)'
C:\dev\Python38
(.venv)
nhanus@nhanus-MOBL MINGW64 ~/dev/git/test_maturin
$ ls /c/dev/Python38
DLLs/  Doc/  include/  Lib/  libs/  LICENSE.txt  NEWS.txt  python.exe*  python3.dll*  python38.dll*  pythonw.exe*  tcl/  Tools/  vcruntime140.dll*
(.venv)
```

Unless there is a linux-style versus windows style path issue ?

**Reply by messense** [reply]
I'm not sure, can you try import it in normal Windows terminal instead of MinGW?

**Reply by NicolasHanus** [reply]
I just tried from a windows CMD. The result is exactly the same :( How could this ImportError be debugged ? The error message is not very verbose :/ I have no clue on what could be the cause of the issue... Thx in advance

**Reply by davidhewitt** [reply]
Can you use `dumpbin.exe` (available in the visual studio developer command prompt) with `dumpbin /symbols`? If you dump symbols for your built `thermal` library as well as for your `python3.dll` we should be able to compare the undefined symbols and hopefully spot whatever is missing.

**Reply by NicolasHanus** [reply]
See below

**Reply by NicolasHanus** [reply]
@davidhewitt @messense I appreciate your help
Any hint of what I could look for in the dumpbin result ? Or any other idea of how I could find out what the dll is looking for at load time ?
Many thanks

**Reply by messense** [reply]
Perhaps you can refer to this article to debug the dll load issue: https://abitofscotland.wordpress.com/2019/12/20/dll-load-failure/

**Reply by NicolasHanus** [reply]
I found and solved the issue. I could see in dependency walker that symbol Py_CompileString was not found in dependent DLL python38.dll
I searched and found [this issue](https://github.com/python/cpython/issues/81814) stating these symbols are indeed missing in Python distribution 3.8.0. That was the one I was using. I installed 3.8.10 instead and it's now working.
Thank you both for your help.

**Reply by NicolasHanus** [reply]
![image](https://user-images.githubusercontent.com/71252267/173065612-c4acd61d-5879-488a-9650-5bdab1db5502.png)

**Comment by NicolasHanus** [comment]
Here is the result on my library.
[log.txt](https://github.com/PyO3/maturin/files/8868510/log.txt)

And here it is for the python3.dll. Strangely the command shows nothing for this one :/
[log.txt](https://github.com/PyO3/maturin/files/8868564/log.txt)

Can you be more specific about what I should be looking for in the first dump ? I see plenty of UNKNOWN or UNDEF but it's not really explicit.

I'm also trying to analyze my library with dependency walker, but so far, the tool is stuck :(

Thx in advance
